Add logic grid puzzle eval

AustinStarnes opened the pull request:
# Thank you for contributing an eval! ♥️

🚨 Please make sure your PR follows these guidelines, __failure to follow the guidelines below will result in the PR being closed automatically__. Note that even if the criteria are met, that does not guarantee the PR will be merged nor GPT-4 access granted. 🚨

__PLEASE READ THIS__:

In order for a PR to be merged, it must fail on GPT-4. We are aware that right now, users do not have access, so you will not be able to tell if the eval fails or not. Please run your eval with GPT-3.5-Turbo, but keep in mind as we run the eval, if GPT-4 gets higher than 90% on the eval, we will likely reject since GPT-4 is already capable of completing the task.

We plan to roll out a way for users submitting evals to see the eval performance on GPT-4 soon. Stay tuned! Until then, you will not be able to see the eval performance on GPT-4. **Starting April 10, the minimum eval count is 15 samples, we hope this makes it easier to create and contribute evals.**

## Eval details 📑
### Eval name
`logic-grid`

### Eval description

The model is prompted with a logic grid word problem with a list of hints, categories, and values for the category. The model is then asked to provide a solution, showing reasoning, in a set of Python tuples that show the matched category values.

<details>
  <summary>See human-friendly example prompt</summary>

```
Using the hints and information provided, solve the following logic grid puzzle.
Show your reasoning concisely, but step-by-step, to arrive at your final answer.
Remember, since this is a grid-based puzzle, each option in every category will be used exactly once.
Since there are 4 values per category, and 3 categories, your final answer will include 4 tuples each of length 3.
 Puzzle Scenario:
A number of new space probes were launched this year. Using only the clues that follow, match each probe to the month it launched and its eventual destination.

Category Options:
  Probes: "Nalini", "Hikaru", "Sarika", and "Sven"
  Destinations: "Ganymede", "Titan", "Enceladus", and "Phobos"
  Months: "January", "February", "March", and "April"

Hints:
  1: The Hikaru launched 2 months before the Nalini.
  2: The probe going to Phobos launched sometime after the Hikaru.
  3: The Sven launched 2 months before the mission going to Ganymede.
  4: Of the probe that launched in April and the mission going to Titan, one is the Sarika and the other is the Nalini.

In your response, format your final answer for each set of linked category values as a python tuple of strings in this order:
("Enceladus", [value from months], [value from probes])
("Ganymede", [value from months], [value from probes])
("Phobos", [value from months], [value from probes])
("Titan", [value from months], [value from probes])
```
</details>

<details>
  <summary>See expected/ideal answer</summary>

```
("Enceladus", "January", "Hikaru")
("Ganymede", "April", "Sarika")
("Phobos", "February", "Sven")
("Titan", "March", "Nalini")
```
</details>

<details>
  <summary>See failure case from GPT-4 (at least from the web UI - I don't have API access to GPT-4)</summary>

```
Step 1: From Hint 3, we know that Hikaru launched 2 months before Nalini. Since the latest month available is April, this means that Nalini cannot have launched in April because there wouldn't be a 2-month gap for Hikaru. So, Nalini must be the probe going to Titan (from Hint 4), and Sarika must be the probe that launched in April.

Step 2: Since Nalini is going to Titan, from Hint 4, it must have launched in February to allow Hikaru to launch in December (2 months before). However, December is not an option, so Hikaru must have launched in January (the earliest month possible) and Nalini in March.

Step 3: The Sven launched 2 months before the mission going to Ganymede (Hint 2). Since Sven cannot be launched in January (as Hikaru is launched in January), it must have launched in February. That means the mission to Ganymede must be in April.

Step 4: The only remaining month is March, and the only remaining probe is Nalini. So Nalini was launched in March.

Step 5: The probe going to Phobos launched sometime after Hikaru (Hint 1). Since Hikaru launched in January, the only remaining option for Phobos is Sarika, which launched in April.

Step 6: The only remaining destination is Enceladus, and the only remaining probe is Hikaru, so Hikaru is going to Enceladus.

The final answer is:
("Enceladus", "January", "Hikaru")
("Ganymede", "April", "Sven")
("Phobos", "April", "Sarika")
("Titan", "March", "Nalini")
```
</details>

### What makes this a useful eval?

This set of evals relies on a lot of *context-ramping* (as I call it) for LLMs to get right - relying on chain-of-thought style logic to 'ramp' into the most likely next-deduction for a multi-parameter logical problem.
These prompts are written as word problems and don't rely on any ASCII or markdown type of table to express the puzzle (though `gpt-3.5-turbo` appears to use markdown tables in it's chain of thought).

This ramping gets expensive and doesn't scale well with the increase of category or values-per-category parameters, and is something that LLMs might benefit from avoiding (either by requiring less of a context-ramp to get to the meat of an answer, or by the surrounding infrastructure/problem agent that tries to use a cheaper ensemble for most of the context ramping, or other strategies).

The reasoning skills required for this kind of <b>constraint satisfaction</b> expressed in natural language word prompts can, in theory, generalize to solving constraints expressed in less structured formats in the wild.

## Criteria for a good eval ✅

Below are some of the criteria we look for in a good eval. In general, we are seeking cases where the model does not do a good job despite being capable of generating a good response (note that there are some things large language models cannot do, so those would not make good evals).

Your eval should be:

- [X] Thematically consistent: The eval should be thematically consistent. We'd like to see a number of prompts all demonstrating some particular failure mode. For example, we can  create an eval on cases where the model fails to reason about the physical world.
- [X] <span title="A GPT-4 failure case from the eval set is included above.">Contains failures where a human can do the task, but either GPT-4 or GPT-3.5-Turbo could not.</span>
- [X] Includes good signal around what is the right behavior. This means either a correct answer for `Basic` evals or the `Fact` Model-graded eval, or an exhaustive rubric for evaluating answers for the `Criteria` Model-graded eval.
- [X] **<span title="15 examples per grid size are included, for 6 different grid sizes, make a total of 90 included examples.">Include at least 15 high quality examples.</span>**

If there is anything else that makes your eval worth including, please document it below.

### Unique eval value

This eval contains varied prompts and grid sizes, 15 examples per grid size, for a total of 90 procedurally generated puzzles.
~This PR also implements (disabled by default) multiline answers and punctuation whitelisting behavior in `FuzzyMatch`, which was required due to the structural nature of the logic grid answers and the behavior of `utils.normalize(...)` which (prior to this PR) did not provide any functionality to interpret answers that span across newline characters `\n` or preserve punctuation where it may be useful (in this case, the model is prompted for the resultant grid associations in python tuple format, so the punctuation in `(",)` is configured to be preserved for these evals).
The new functionality is, again, disabled by default, such that existing evals are not affected.~

## Eval structure 🏗️

Your eval should
- [X] Check that your data is in `evals/registry/data/{name}`
- [X] Check that your yaml is registered at `evals/registry/evals/{name}.yaml`
- [X] Ensure you have the right to use the data you submit via this eval

(For now, we will only be approving evals that use one of the existing eval classes. You may still write custom eval classes for your own cases, and we may consider merging them in the future.)

## Final checklist 👀

### Submission agreement

By contributing to Evals, you are agreeing to make your evaluation logic and data under the same MIT license as this repository. You must have adequate rights to upload any data used in an Eval. OpenAI reserves the right to use this data in future service improvements to our product. Contributions to OpenAI Evals will be subject to our usual Usage Policies (https://platform.openai.com/docs/usage-policies).

- [X] I agree that my submission will be made available under an MIT license and complies with OpenAI's usage policies.

### Email address validation

If your submission is accepted, we will be granting GPT-4 access to a limited number of contributors. Access will be given to the email address associated with the merged pull request.

- [X] I acknowledge that GPT-4 access will only be granted, if applicable, to the email address used for my merged pull request.

### Limited availability acknowledgement

We know that you might be excited to contribute to OpenAI's mission, help improve our models, and gain access to GPT-4. However, due to the requirements mentioned above and high volume of submissions, we will not be able to accept all submissions and thus not grant everyone who opens a PR GPT-4 access. We know this is disappointing, but we hope to set the right expectation before you open this PR.

- [X] I understand that opening a PR, even if it meets the requirements above, does not guarantee the PR will be merged nor GPT-4 access granted.

### Submit eval

- [X] I have filled out all required fields in the evals PR form
- [X] (Ignore if not submitting code) I have run `pip install pre-commit; pre-commit install` and have verified that `black`, `isort`, and `autoflake` are running when I commit and push

Failure to fill out all required fields will result in the PR being closed.

### Eval JSON data 

Since we are using Git LFS, we are asking eval submitters to add in as many Eval Samples (at least 5) from their contribution here:

<details>
  <summary>View evals in JSON</summary>

  ### Eval
  ```jsonl
{"input":[{"role":"system","content":"Using the hints and information provided, solve the following logic grid puzzle.\nShow your reasoning concisely, but step-by-step, to arrive at your final answer.\nRemember, since this is a grid-based puzzle, each option in every category will be used exactly once.\nSince there are 4 values per category, and 3 categories, your final answer will include 4 tuples each of length 3.\n"},{"role":"system","content":"Puzzle Scenario:\nA number of new space probes were launched this year. Using only the clues that follow, match each probe to the month it launched and its eventual destination.\n\nCategory Options:\n  Probes: \"Nalini\", \"Hikaru\", \"Sarika\", and \"Sven\"\n  Destinations: \"Ganymede\", \"Titan\", \"Enceladus\", and \"Phobos\"\n  Months: \"January\", \"February\", \"March\", and \"April\"\n\nHints:\n  1: The Hikaru launched 2 months before the Nalini.\n  2: The probe going to Phobos launched sometime after the Hikaru.\n  3: The Sven launched 2 months before the mission going to Ganymede.\n  4: Of the probe that launched in April and the mission going to Titan, one is the Sarika and the other is the Nalini.\n\nIn your response, format your final answer for each set of linked category values as a python tuple of strings in this order:\n(\"Enceladus\", [value from months], [value from probes])\n(\"Ganymede\", [value from months], [value from probes])\n(\"Phobos\", [value from months], [value from probes])\n(\"Titan\", [value from months], [value from probes])"}],"ideal":"(\"Enceladus\", \"January\", \"Hikaru\")\n(\"Ganymede\", \"April\", \"Sarika\")\n(\"Phobos\", \"February\", \"Sven\")\n(\"Titan\", \"March\", \"Nalini\")"}
{"input":[{"role":"system","content":"Using the hints and information provided, solve the following logic grid puzzle.\nShow your reasoning concisely, but step-by-step, to arrive at your final answer.\nRemember, since this is a grid-based puzzle, each option in every category will be used exactly once.\nSince there are 4 values per category, and 3 categories, your final answer will include 4 tuples each of length 3.\n"},{"role":"system","content":"Puzzle Scenario:\nThe Monterey Bay Aquarium has a new exhibit featuring several dolphin skeletons. The aquarium asked local elementary school children to name each dolphin and match it to its species, length, and the year in which it was donated to the aquarium. Use the given clues to solve the puzzle.\n\nCategory Options:\n  Dolphin: \"Atlantic white-sided dolphin\", \"Pacific white-sided dolphin\", \"common dolphin\", and \"bottlenose dolphin\"\n  Names: \"Momo\", \"Simon\", \"Tychus\", and \"Marlo\"\n  Lengths: \"21 ft\", \"23 ft\", \"25 ft\", and \"27 ft\"\n\nHints:\n  1: Momo is 4 feet shorter than Tychus.\n  2: The common dolphin is somewhat shorter than the bottlenose dolphin.\n  3: The 21 ft long exhibit is either the Atlantic white-sided dolphin or Simon.\n  4: Marlo is either the 23 ft long exhibit or the Atlantic white-sided dolphin.\n  5: Of Simon and the Pacific white-sided dolphin, one is 27 ft long and the other is 21 ft long.\n\nIn your response, format your final answer for each set of linked category values as a python tuple of strings in this order:\n(\"Atlantic white-sided dolphin\", [value from lengths], [value from names])\n(\"Pacific white-sided dolphin\", [value from lengths], [value from names])\n(\"bottlenose dolphin\", [value from lengths], [value from names])\n(\"common dolphin\", [value from lengths], [value from names])"}],"ideal":"(\"Atlantic white-sided dolphin\", \"25 ft\", \"Marlo\")\n(\"Pacific white-sided dolphin\", \"27 ft\", \"Tychus\")\n(\"bottlenose dolphin\", \"23 ft\", \"Momo\")\n(\"common dolphin\", \"21 ft\", \"Simon\")"}
{"input":[{"role":"system","content":"Using the hints and information provided, solve the following logic grid puzzle.\nShow your reasoning concisely, but step-by-step, to arrive at your final answer.\nRemember, since this is a grid-based puzzle, each option in every category will be used exactly once.\nSince there are 4 values per category, and 3 categories, your final answer will include 4 tuples each of length 3.\n"},{"role":"system","content":"Puzzle Scenario:\nHillcrest County Hospital released a press release today to announce its annual list of the first babies born immediately after the start of the new year. Using only the clues below, match each baby to its family time of birth, and determine the room number in which it was delivered.\n\nCategory Options:\n  Names: \"Lydia\", \"Naomi\", \"Katherine\", and \"Rebecca\"\n  Families: \"McIntyre\", \"Gonzalez\", \"Hudson\", and \"O'Connor\"\n  Times: \"12:01am\", \"12:04am\", \"12:07am\", and \"12:10am\"\n\nHints:\n  1: Naomi was born 3 minutes after Rebecca.\n  2: The Hudsons' child was born 6 minutes after the O'Connors' child.\n  3: The baby born at 12:04am was either Rebecca or the McIntyres' child.\n  4: Of Katherine and the baby born at 12:01am, one was the O'Connors' and the other was the Gonzalezes'.\n\nIn your response, format your final answer for each set of linked category values as a python tuple of strings in this order:\n(\"Gonzalez\", [value from names], [value from times])\n(\"Hudson\", [value from names], [value from times])\n(\"McIntyre\", [value from names], [value from times])\n(\"O'Connor\", [value from names], [value from times])"}],"ideal":"(\"Gonzalez\", \"Katherine\", \"12:10am\")\n(\"Hudson\", \"Lydia\", \"12:07am\")\n(\"McIntyre\", \"Naomi\", \"12:04am\")\n(\"O'Connor\", \"Rebecca\", \"12:01am\")"}
{"input":[{"role":"system","content":"Using the hints and information provided, solve the following logic grid puzzle.\nShow your reasoning concisely, but step-by-step, to arrive at your final answer.\nRemember, since this is a grid-based puzzle, each option in every category will be used exactly once.\nSince there are 4 values per category, and 3 categories, your final answer will include 4 tuples each of length 3.\n"},{"role":"system","content":"Puzzle Scenario:\nMedisynth Therapeutics has spent decades scouring the world's rainforests for new sources of medical drugs, and this year a number of these new drugs have been officially approved by the FDA. Using only the clues below, match each drug to the condition it treats, the month it was approved, and the source from which its main ingredient is derived.\n\nCategory Options:\n  Drugs: \"Mystolam\", \"Soludox\", \"Aurexil\", and \"Synodil\"\n  Conditions: \"rheumatism\", \"chikungunya\", \"hyperglycemia\", and \"encephalitis\"\n  Months: \"January\", \"February\", \"March\", and \"April\"\n\nHints:\n  1: The medicine that treats hyperglycemia was approved sometime after Synodil.\n  2: The medicine that treats hyperglycemia was approved 1 month before the medicine that treats rheumatism.\n  3: Of Mystolam and the pharmaceutical approved in April, one treats rheumatism and the other treats encephalitis.\n  4: Of the Aurexil approved in February and the drug that treats chikungunya, one is Soludox and the other is Synodil.\n\nIn your response, format your final answer for each set of linked category values as a python tuple of strings in this order:\n(\"chikungunya\", [value from drugs], [value from months])\n(\"encephalitis\", [value from drugs], [value from months])\n(\"hyperglycemia\", [value from drugs], [value from months])\n(\"rheumatism\", [value from drugs], [value from months])"}],"ideal":"(\"chikungunya\", \"Synodil\", \"January\")\n(\"encephalitis\", \"Soludox\", \"April\")\n(\"hyperglycemia\", \"Aurexil\", \"February\")\n(\"rheumatism\", \"Mystolam\", \"March\")"}
{"input":[{"role":"system","content":"Using the hints and information provided, solve the following logic grid puzzle.\nShow your reasoning concisely, but step-by-step, to arrive at your final answer.\nRemember, since this is a grid-based puzzle, each option in every category will be used exactly once.\nSince there are 4 values per category, and 3 categories, your final answer will include 4 tuples each of length 3.\n"},{"role":"system","content":"Puzzle Scenario:\nA group of friends has gotten together for dinner, and just for fun they're comparing their \"social networking\" stats on Facebook, Twitter and LinkedIn. Using only the clues that follow, determine the number of friends (Facebook), followers (Twitter) and connections (LinkedIn) each person has.\n\nCategory Options:\n  Names: \"Avery\", \"Evan\", \"Natalia\", and \"Camila\"\n  Linkedin: \"54\", \"59\", \"64\", and \"68\"\n  Facebook: \"120\", \"130\", \"140\", and \"150\"\n\nHints:\n  1: Avery has 68 LinkedIn connections.\n  2: Natalia has 10 fewer Facebook friends than the person with 68 LinkedIn connections.\n  3: The person with 140 Facebook friends is either the one with 54 LinkedIn connections or Evan.\n  4: The one with 54 LinkedIn connections has 10 fewer Facebook friends than the one with 64 LinkedIn connections.\n\nIn your response, format your final answer for each set of linked category values as a python tuple of strings in this order:\n(\"120\", [value from LinkedIn], [value from names])\n(\"130\", [value from LinkedIn], [value from names])\n(\"140\", [value from LinkedIn], [value from names])\n(\"150\", [value from LinkedIn], [value from names])"}],"ideal":"(\"120\", \"59\", \"Natalia\")\n(\"130\", \"68\", \"Avery\")\n(\"140\", \"54\", \"Camila\")\n(\"150\", \"64\", \"Evan\")"}
  ```
</details>

luqman-openai reviewed (changes requested):
Thanks for submitting this eval! I would like to recommend some changes.

- It seems you are modifying the `fuzzy_match.py` and the `utils.py`, it is not recommended to modify the code base in the eval PR. If the changes are really necessary open a separate PR.

- What were the results of the `FuzzyMatch` evaluator before the modification? 
-- If the results are reasonable then we can go with it without the modifications.
-- Alternately, you can make use of model graded eval with `fact` spec

- It seems you have registered multiple evals for the different grid sizes. It is recommended to merge the sample files in to a single one, the prompt instructions are clear. It should not cause any issues. Creating different evals for different grid sizes will be considered duplicates of each other.
-- `evals/registry/eval_sets/logic-grid.yaml` can be removed after the above mentioned changes.

- The naming convention for evals should be in the form `<eval_name>.<split>.<version>`.
-- `<eval_name>` is the eval name, used to group evals whose scores are comparable. 
-- `<split>` is the data split, used to further group evals that are under the same <base_eval>. E.g., `val`, `test`, or `dev` for testing. 
-- `<version>` is the version of the eval, which can be any descriptive text you'd like to use (though it's best if it does not contain `.`).

We would love to review your PR again after the suggested changes are incorporated.

AustinStarnes commented:
Hi @luqman-openai, thank you for the thoughtful and thorough feedback. I appreciate the guidance and your patience with my improvisations to make such a complicated eval possible.

### Changes to FuzzyMatch and normalize are reverted, at least for this PR
Firstly, I was pleased (and relieved) to see that the main branch has incorporated a fix (#1026) that covers the principal reason why I found it necessary to modify `FuzzyMatch`; and that was that only the first line of a completion was being retained (let alone compared) in `FuzzyMatch`. #1026 greatly mitigates any need for a `multiline` option in the `normalize` function, and therefore I have removed it.

With that change in mind, the 2nd enhancement I had incorporated, which allowed a whitelist of punctuation to be preserved in a FuzzyMatch, is much less important. I therefore found it reasonable to remove the `preserve_punct` option of the `normalize` function as well, which allowed me to revert the `**kwargs` passthrough in other parts of the code as well.

### Changes in response to your guidance and feedback
- As mentioned above, `FuzzyMatch` and `normalize` are no longer changed in this PR.
  - The multiline option in `normalize` is no longer necessary and has been removed, as `normalize` was changed to not discard lines in #1026 
  - The preserve_punct option in `normalize` would help, in the case of this eval, determine where the newlines occurred, as `normalize` eventually splits the completion by whitespace, before joining the split with simple spaces. This means that newlines should not have significance in the accuracy of an answer for `FuzzyMatch`. If punctuation is preserved, then the newlines can be discarded without issue, as this eval prompts the model to format its answer in the form of python tuples which are encapsulated with `()`. **However, I have removed the preserve_punct option for this PR anyway, per the guidance that necessary modifications be proposed in a dedicated PR.** Correspondingly, this change is less important now that `FuzzyMatch` can handle multiple lines of completion. **I only plan to reimplement this in a separate PR if given guidance here that the change is consistent with the general purpose and design of `FuzzyMatch`/`normalize`,**, but I am happy to do so in that case!
  - The change to respect `max_tokens` in `FuzzyMatch` completions has been ported to the dedicated PR #1076.
    See related #989 and #1076. and my detailed explanation below.
- All logic grid puzzles are provided under a single topological eval rather than one eval per grid size.
**NOTE** the main reason they were separated to begin with was due to the speculation that larger grids would require proportionally, perhaps exponentially more tokens to solve. However, `max_tokens`, was only configurable per-eval, and only after my changes to `FuzzyMatch`. 
- The logic grid eval set has been removed.
- The eval now follows the described naming convention, though it assigns every puzzle to the `dev` split. It is not clear to me how we should determine whether to separate certain eval input into `test` and `val` splits.

### It is important to note that GPT-4 will perform with rougly 0.00% accuracy without the minimal changes introduced in #1076
Like I mention in #1076, even GPT-4 struggles to arrive at any good answer without talking through the hints of the puzzle before making final assignments of categories and values from the puzzle.
For example, here is some test output from GPT-4:
```
Let's start solving the puzzle step by step using the hints provided.

Hint 1: The tortoise was lemon.
This directly gives us the color of the tortoise balloon.

Hint 2: The poodle cost $6.00.
This directly gives us the price of the poodle balloon.

Hint 3: The balloon Hector bought cost less than the fuchsia balloon.
This means Hector didn't buy the fuchsia balloon.

Hint 4: The champagne balloon cost 1 dollar less
```
![image](https://github.com/openai/evals/assets/103542593/af3d0b9f-d90b-4058-a084-aab9134a22f7)

You can see that `FuzzyMatch`'s hardcoded `max_tokens` length 100 stopped the completion from even providing a grid solution, let alone reason out the correct one.

Before creating this eval, I experimented extensively to determine the optimal phrasing and generation strategies to give GPT-3.5 and GPT-4 the best shot at answering successfully.
Prompt strategies for these puzzles that coax an answer out immediately, e.g. "In your response, provide ONLY the solution to the puzzle grid...", resulted in **worse** performance (lower accuracy, though GPT-3.5 scored a perfect 0.00% no matter what I did, even for the smallest grid size).

**Therefore, in order for a completion function on par with GPT-4 to have a nonzero accuracy to even register improvement upon, I recommend exposing control over `FuzzyMatch` completion token amounts to the eval configuration.** (Hence my minimal implementation of this in #1076, and previously in this very PR before I reverted it)

### Other comments
Most of my changes were improvisations to make `FuzzyMatch` meet my needs without changing its existing behavior *or* to anticipate heavy/expensive processing burden required to run this eval.
For example, separating the eval by grid size meant I could specify the smaller logic grid puzzles to have a smaller `max_tokens` for the completion function, making the eval less expensive to run with a model such as `GPT-4`.
I felt at the time the cost of the model justified such a distinction, but understand now based on your guidance that you'd probably recommend managing cost by other means, such as not running the eval with `GPT-4`.
Likewise, I avoided Model-Graded evals because I found that they were prohibitively expensive even without the expense of the `max_tokens` necessary for the completion functions I tried to get correct answers at all. ModelGraded had indeed been my first choice due to the multiline issue with `FuzzyMatch`.

luqman-openai reviewed (approved):
Thanks for the for the updates and the detailed explanation. I have shared your fix, https://github.com/openai/evals/pull/1076 with other team members to have a look at it. I appreciate your time and effort to investigate the issue.

The eval looks good to me! I am approving this PR.

You should see GPT-4 API access enabled in your account in the few days.
